OCR engines selection ADR (#15708)
JabRef/jabref: docs/decisions/0056-OCR-engine-selection.md

- Good, because MIT license is fully compatible.
- Bad, because output is Markdown text, not a pre-built searchable PDF a text layer embedding step is still required.
- Bad, because ML model weights (~1–2 GB) are downloaded on first use.
- Bad, because it requires Python 3.x.

### olmOCR

Homepage: <https://github.com/allenai/olmocr>
License: [Apache 2.0](https://github.com/allenai/olmocr?tab=Apache-2.0-1-ov-file#readme)
Integration: Subprocess via `ProcessBuilder`

olmOCR is an OCR toolkit from the Allen Institute for AI (AI2), specifically trained and optimized for academic PDF documents using a vision-language model architecture.

- Good, because it is specifically optimized for academic PDFs the primary document type JabRef users work with.
- Good, because Apache 2.0 license is fully compatible.
- Good, because subprocess integration pattern is identical to OCRmyPDF and Docling.
- Bad, because GPU hardware is required for reasonable performance.
- Bad, because model weights are downloaded on first use.
- Bad, because it is newer and less tested than OCRmyPDF or Tesseract.

### PaddleOCR-VL via llama.cpp

> Homepage:<https://huggingface.co/PaddlePaddle/PaddleOCR-VL>

Reviewer: ```suggestion
Homepage:<https://github.com/ggml-org/llama.cpp>
License: [MIT](https://github.com/ggml-org/llama.cpp#MIT-1-ov-file)
Integration: HTTP API via locally running llama.cpp server

Llama.cpp features LLM inference in C/C++, which includes Text, Audio and Vision (OCR) based language models, as well as embedding models.

- Good, because MIT license is fully compatible.
- Good, because the vast majority of model makers release models in Llama.cpp's GGUF format.
- Good, because a great amount of inference backends are supported (CPU, CUDA, Vulkan, SYCL and others), which means compatibility with a wide range of hardware. Depending on hardware and backend, inference can be MUCH faster than pure CPU based inference. Backend support status: [Table of supported operands](https://github.com/ggml-org/llama.cpp/blob/master/docs/ops.md).
- Good, because model makers regularly release new models and llama.cpp now supports multiple OCR models.
- Good, because users can swap models based on their needs.
- Good, because the HTTP API integration pattern is reusable for other remote OCR engines.
- Good, because text, audio, vision, embeddings and OCR inference is combined in a single framework.
- Good, because it is one of the major frameworks. Has a large community; many stars on GitHub.
- Bad, because the framework is a "can do it all" and not a dedicated OCR engine/framework.
- Bad, because only a limited number of OCR models are supported at present.
- Bad, because requires a separate framework for running a VLM.
- Bad, because java bindings are not well supported.
- Bad, because either the user is forced to separately install and run a llama.cpp server instance or JabRef developers have to maintain bindings to bundle it natively into JabRef.
- Bad, because size of model weights can require users to download several GB depending on model variant.

### PaddleOCR-VL via llama.cpp

llama.cpp integration: <https://github.com/ggml-org/llama.cpp/pull/16701>
License: [Apache 2.0](https://huggingface.co/datasets/choosealicense/licenses/blob/main/markdown/apache-2.0.md)
Integration: HTTP API via locally running llama.cpp server

PaddleOCR-VL is a vision-language model achieving state-of-the-art accuracy on OCR benchmarks as of October 2025. llama.cpp PR 16701 adds support for running PaddleOCR-VL locally via llama.cpp's server, which exposes an HTTP API. For JabRef, integration is as a `RemoteOcrEngine` sending requests to a user-provisioned llama.cpp instance.

- Good, because Apache 2.0 license is fully compatible.
- Good, because users can simply download and run the model with a preconfigured fixed system prompt.
- Bad, because llama.cpp PR 16701 is recent and less tested than OCRmyPDF or Tesseract.
- Bad, because it has all the typical architectural drawbacks of Vision Language Models: confabulation and hallucination.
- Bad, because there is no guarantee PaddleOCR will support successor models in llama.cpp
```

> # ADR for OCR engine selection

Reviewer 2: Please follow the [ADR template](https://github.com/JabRef/jabref/blob/main/docs/decisions/adr-template.md?plain=1)
Example:
https://github.com/JabRef/jabref/blob/cc371f76149baba1104f94c191de39760fba2015/docs/decisions/0014-separate-URL-creation-to-enable-proper-logging.md?plain=1#L1-L5

Apache Tika is not selected as a primary OCR path because it abstracts away fine-grained settings and does not improve on OCRmyPDF for scanned documents. Surya and Kreuzberg are excluded on license grounds. SimpleHTR and deepdoctection are documented for future consideration but are out of scope for this project phase.

### Consequences

- Good, because OCRmyPDF handles the embedding the text layer automatically.
- Good, because OCRmyPDF will allow users to use different engines through the plugins it offers.
- Good, because OCRmyPDF's `--skip-text` flag handles partially searchable PDFs by skipping pages that already have a valid text layer.
- Bad, because OCRmyPDF requires Python 3.x to be installed on the user's machine, which is an additional step for non-technical users.
- Bad, because Tess4J's JNA native library loading remains fragile.
- Bad, because Docling and olmOCR download machine learning model weights (1–2 GB) on first use, which must be an explicit user opt-in and cannot happen silently.

### Confirmation

The integration will be confirmed by a `DocumentReaderTest` that verifies:

> 1. `PDFTextStripper` returns empty content for a scanned PDFs before performing OCR (established in PR [#15428](https://github.com/JabRef/jabref/pull/15428)).

Reviewer 2: ```suggestion
1. `PDFTextStripper` returns empty content for a scanned PDF before performing OCR (established in PR [#15428](https://github.com/JabRef/jabref/pull/15428)).
```

Author: I accidentally used an outdated version for this section.

Now, the ADR is up to date with the latest version, including the edits and comments you left.

---
parent: Decision Records
nav_order: 56
---
# ADR for OCR engine selection

## Context and Problem Statement

JabRef requires an OCR engine to extract text from scanned academic PDFs, especially historic and old documents. Academic documents present unique challenges: mathematical notation, multiple languages, complex layouts, tables, and mixed handwritten/printed content.

Which OCR engines should JabRef support to serve its academic user base?

> ---

Reviewer 3: We don't use `---` in ADR, so needs to be removed

> ## Decision Drivers

Reviewer 3: I think it's better to use `should` wording instead of `must`, `must` is too strong

> - GPL and non-open-source are not acceptable (Online services are okay).

Reviewer 3: ```suggestion
- GPL and non-open-source are not acceptable (with exception to online services).
```

- Bad, because Docling and olmOCR download machine learning model weights (1–2 GB) on first use, which must be an explicit user opt-in and cannot happen silently.

### Confirmation

The integration will be confirmed by a `DocumentReaderTest` that verifies:

1. `PDFTextStripper` returns empty content for scanned PDFs before performing OCR (established in PR [#15428](https://github.com/JabRef/jabref/pull/15428)) for only the PDFs that are fully scanned and have no text at all (some PDFs have partially selectable text).
2. After OCR processing, `PDFTextStripper` returns more selectable content than before from the same file.

---

## Pros and Cons of the Options

### OCRmyPDF

Homepage: [https://github.com/ocrmypdf/OCRmyPDF](https://github.com/ocrmypdf/OCRmyPDF)
License: [MPL-2.0](https://github.com/ocrmypdf/OCRmyPDF#MPL-2.0-1-ov-file)
Integration: Subprocess via `ProcessBuilder`

OCRmyPDF is a Python command-line tool that takes a scanned PDF, OCRs each page using Tesseract internally, and outputs a new PDF with an invisible text layer. The output PDF is immediately readable by Lucene, requiring no further PDF production.

> - Good, because both OCR extraction and text layer embedding are handled in a single subprocess call no separate step needed for embedding the text layer into the PDF.

Reviewer 3: ```suggestion
- Good, because both OCR extraction and text layer embedding are handled in a single subprocess call, and no separate step is needed for embedding the text layer into the PDF.
```

- Good, because it uses hOCR to position words producing accurately located invisible text, enabling correct copy-paste and PDF viewer search highlighting.
- Good, because `--skip-text` correctly handles partially searchable PDFs without re-OCRing already-good pages.

> - Good, because it [plugins](https://github.com/ocrmypdf/OCRmyPDF#plugins) that uses different engines from Tesseract like [OCRmyPDF-PaddleOCR](https://github.com/clefru/ocrmypdf-paddleocr) that replaces the standard Tesseract OCR engine with PaddleOCR.

Reviewer 3: ```suggestion
- Good, because it has [plugins](https://github.com/ocrmypdf/OCRmyPDF#plugins) that uses different engines from Tesseract like [OCRmyPDF-PaddleOCR](https://github.com/clefru/ocrmypdf-paddleocr) that replaces the standard Tesseract OCR engine with PaddleOCR.
```

- Good, because MPL-2.0 is compatible with JabRef's MIT license for subprocess invocation with no source modification.
- Good, because it is available via `pip`, `apt`, `brew`, and standalone installers on all three supported platforms.
- Bad, because it requires Python 3.x and pip to be available on the user's machine.
- Bad, because subprocess launch takes some time, which is acceptable for background tasks.

### Tesseract via Tess4J

> Homepage: [https://github.com/nguyenq/tess4j](https://github.com/nguyenq/tess4j)

Reviewer 3: ```suggestion
Tess4j homepage: [https://github.com/nguyenq/tess4j](https://github.com/nguyenq/tess4j)
```

Integration: Java API via JNA (Java Native Access)

Tess4J is a Java wrapper for the Tesseract C++ OCR library. It calls `tesseract.doOCR(file)` directly from Java using JNA, returning extracted text as a `.txt` file. When used as the OCR engine, a separate step is required to embed the extracted text as an invisible layer. Correct multi-page embedding requires per-page Tesseract calls with hOCR output to obtain word coordinates instead of words distribution method as in [SearchablePdfCreator.java](https://github.com/JabRef/jabref/pull/13313/changes#diff-8cb8cef63fa6d8197687f74c560f9ab65a39a4fc566f273af98ef12dd578e25a).

- Good, because it requires no Python installation pure Java with a native library dependency.
- Good, because Apache 2.0 license is fully compatible with JabRef's MIT license.
- Good, because Tesseract supports 100+ languages via downloadable `.traineddata` files.
- Bad, because `module-info.java` requires explicit JNA module access declarations.
- Bad, because correct multi-page text layer embedding requires per-page hOCR processing, adding significant implementation complexity compared to OCRmyPDF.

> - Bad, because tessdata path discovery differs across Windows, macOS, and Linux.

Reviewer 3: ```suggestion
- Bad, because `tessdata` path discovery differs across Windows, macOS, and Linux.
```

- Bad, because Tika abstracts away fine-grained OCR settings, making expert user configuration harder than with direct Tess4J or OCRmyPDF.
- Bad, because Tika internally still calls a system-installed Tesseract binary no installation advantage over direct Tess4J use.
- Bad, because it is not extendable. There is no common OCR interface.

### Docling

Homepage: [https://github.com/docling-project/docling](https://github.com/docling-project/docling)
License: [MIT](https://github.com/docling-project/docling?tab=MIT-1-ov-file#readme)
Integration: Subprocess via `ProcessBuilder`

Docling is a Python document intelligence library from IBM Research. It performs full document understanding: layout analysis, reading order detection, table structure recognition, and OCR via Tesseract or EasyOCR. It outputs clean Markdown or structured JSON.

- Good, because it produces significantly better output quality for complex academic documents: correct multi-column reading order, table recognition, and formula identification.
- Good, because MIT license is fully compatible.

> - Good, because 53k+ GitHub stars indicate strong community adoption and longevity.

Reviewer 3: I agree that we prefer more "popular" engines, so it means they are maintained and updated. But I think it's better not to put it into the `Good` consequences.

On the other hand, if we have some unmaintained OCR engine/tool, then it's better to add a `Bad` consequence

Author: I agree, I think this is actually better.

- Bad, because output is Markdown text, not a pre-built searchable PDF a text layer embedding step is still required.
- Bad, because ML model weights (~1–2 GB) are downloaded on first use.
- Bad, because it requires Python 3.x.

### olmOCR

> Homepage: [https://github.com/allenai/olmocr](https://github.com/allenai/olmocr)

Reviewer 3: You can use `<` and `>` to put URL as is:

```suggestion
Homepage: <https://github.com/allenai/olmocr>
```

Could you please apply this to other links as well?

License: [Apache 2.0](https://github.com/allenai/olmocr?tab=Apache-2.0-1-ov-file#readme)
Integration: Subprocess via `ProcessBuilder`

olmOCR is an OCR toolkit from the Allen Institute for AI (AI2), specifically trained and optimized for academic PDF documents using a vision-language model architecture.

- Good, because it is specifically optimized for academic PDFs the primary document type JabRef users work with.
- Good, because Apache 2.0 license is fully compatible.
- Good, because subprocess integration pattern is identical to OCRmyPDF and Docling.
- Bad, because GPU hardware is required for reasonable performance.

> - Bad, because model weights are downloaded on first use.

Reviewer 3: But why this is a bad reason?

Author: The huge size of the models should be downloaded by the users to be able to use the engine, too many GBs.

Reviewer: Are they downloaded automatically? It is not possible to prevent it from downloading?

Author: In order to use them, users have to download the models, but we can't download them without taking permission from users.

Reviewer 3: Do you mean there is no way to control the downloading of the model? And the first time someone uses it, it is downloaded?

Author: I am not sure actually if we can control the downloading of the models or not, but to let the users use olmOCR the engines have to be installed on the users machines.

Reviewer 3: Okay then I think we should elaborate more on this outcome. Something like "...., which can lead to unwanted downloads"

Author: Anyway, according to the decision outcome olmOCR will be a stretch goal so we can have a discussion then and don't focus on it right now.

- Bad, because it is newer and less tested than OCRmyPDF or Tesseract.

> ### PaddleOCR-VL via llama.cpp

Reviewer 3: But is it possible to run it via llama.cpp? Because if not, then I think this option should be removed

Author: I don't know actually, need to investigate this more. I only saw the PR in the resources that enabled PaddleOCR through llama.cpp.

I think whether it is possible or not, we can exclude PaddleOCR and use it through the OCRmyPDF plugin that uses the PaddleOCR engine instead of tesseract.

Reviewer: It is not possible to run PaddleOCR engine via llama.cpp. Only that one particular vision model.

Homepage: [https://huggingface.co/PaddlePaddle/PaddleOCR-VL](https://huggingface.co/PaddlePaddle/PaddleOCR-VL)
llama.cpp integration: [https://github.com/ggml-org/llama.cpp/pull/16701](https://github.com/ggml-org/llama.cpp/pull/16701)
License: [Apache 2.0](https://huggingface.co/datasets/choosealicense/licenses/blob/main/markdown/apache-2.0.md)
Integration: HTTP API via locally running llama.cpp server

PaddleOCR-VL is a vision-language model achieving state-of-the-art accuracy on OCR benchmarks as of October 2025. llama.cpp PR 16701 adds support for running PaddleOCR-VL locally via llama.cpp's server, which exposes an HTTP API. For JabRef, integration is as a `RemoteOcrEngine` sending requests to a user-provisioned llama.cpp instance.

> - Good, because depending on the OCR model, there can be state-of-the-art accuracy on diverse document types and scripts including non-Latin. Model makers regularly release new models.

Reviewer 3: I think it's better to formulate this as "customizability"

[https://github.com/ggml-org/llama.cpp/blob/master/docs/ops.md](https://github.com/ggml-org/llama.cpp/blob/master/docs/ops.md)

### TurboOCR

Homepage: [https://github.com/aiptimizer/TurboOCR](https://github.com/aiptimizer/TurboOCR)
License: [MIT](https://github.com/aiptimizer/TurboOCR?tab=MIT-1-ov-file#readme)
Integration: HTTP API

TurboOCR is a C++/CUDA server wrapping PaddleOCR's PP-OCRv5 with TensorRT FP16 acceleration. It achieves 270 images/second on FUNSD benchmarks and exposes HTTP and gRPC APIs returning JSON with per-word bounding boxes. It represents the institutional/GPU-infrastructure tier of OCR integration.

- Good, because MIT license is fully compatible.
- Good, because JSON response includes bounding boxes for accurately positioned text layer embedding.
- Good, because the HTTP adapter pattern it establishes covers TurboOCR, Google Cloud Vision, Azure Computer Vision, and any other HTTP-based OCR service a single `RemoteOcrEngine` implementation serves all.
- Bad, because to date, it requires NVIDIA GPU hardware, thereby excluding a wide range of users, making it unsuitable as a default engine.

> - Bad, because it raises privacy concerns when documents contain unpublished research materials.

Reviewer 3: I couldn't find any issues about this consequence, so propbably it should be removed

Author: If some researchers have, for example, some private research, I think they will not be comfortable with sharing them to a cloud service OCR engine.

Reviewer 3: Yes, but if someones decides to use an online service, then the person should take the responsibility of all riscs.

+ What if this server runs locally? Then it won't have any privacy issues

Author: > Yes, but if someones decides to use an online service, then the person should take the responsibility of all riscs.
 
Yeah, for sure, we can raise a note for users who want to use cloud engines.

> * What if this server runs locally? Then it won't have any privacy issues

I think so, but we need to make sure that we can run on a local server first.

Reviewer: ```suggestion
- Bad, because as at the date of writing only supports remote OCR services, which raises privacy concerns, when documents contain unpublished research materials.
```

If TurboOCR only supports (local only) offline mode, this is a no issue. If they support online services, we can do something similar like we did with JabRef's AI features, when first opening the AI summary or AI chat tabs: Warn and tell users that all responsibility lies with them, if they send data to remote services.

Author: > If TurboOCR only supports (local only) offline mode, this is a no issue. If they support online services, we can do something similar like we did when activating JabRef's AI features: Warn and tell users that all responsibility lies with them, if they send data to remote services.

Ok, perfect. We have to double-check that it can be used locally, but I think it's not the time for it now.

- Bad, because to date, the project only supports a few select language family bundles (Latin, Chinese, Greek, Korean, Arabic, eslav, thai).

### deepdoctection

Homepage: [https://github.com/deepdoctection/deepdoctection](https://github.com/deepdoctection/deepdoctection)
License: [Apache 2.0](https://github.com/deepdoctection/deepdoctection?tab=Apache-2.0-1-ov-file#readme)
Integration: Subprocess via `ProcessBuilder`

deepdoctection is a Python Document AI orchestration framework coordinating layout analysis (Detectron2/Transformers), OCR (Tesseract, DocTr, AWS Textract), table structure recognition, and token classification. It produces structured output capturing the full document hierarchy and is the most comprehensive document understanding pipeline among all evaluated options.

- Good, because Apache 2.0 license is fully compatible.
- Good, because it represents the state of the art in document understanding pipelines and serves as the reference architecture for what a comprehensive future JabRef OCR pipeline could look like.

> - Bad, because installation requires Detectron2, PyTorch, and multiple model weights..

Reviewer 3: ```suggestion
- Bad, because installation requires Detectron2, PyTorch, and multiple model weights.
```

- Bad, because the installation complexity makes it impractical to document as a user-installable engine for most JabRef users.

> deepdoctection is not selected for this project phase but is retained as the reference architecture for future comprehensive document understanding.

Reviewer 3: I think this could be moved to the `Decision outcome`

### SimpleHTR

Homepage: [https://github.com/githubharald/SimpleHTR](https://github.com/githubharald/SimpleHTR)
License: [MIT](https://github.com/githubharald/SimpleHTR?tab=MIT-1-ov-file#readme)
Integration: Subprocess via `ProcessBuilder`

SimpleHTR is a TensorFlow model trained specifically for handwritten text recognition on historical documents. It addresses the use case that all other evaluated engines handle poorly: handwriting in manuscripts, letters, and archival records.

- Good, because MIT license is fully compatible.
- Good, because it uniquely addresses handwritten historical documents a genuine need for historians and archivists using JabRef.
- Bad, because it requires TensorFlow, making it a large dependency for a narrow use case.
- Bad, because it produces raw text output without bounding boxes, requiring simple page-level text embedding without word-level position accuracy.

> - Bad, because it is appropriate only for a small subset of JabRef users.

Reviewer 3: I think this should be replaced with that it focuses on the handwritten text, but not generally applicable to scientific articles
